feat: introducing post updated worker

rebelchris opened the pull request:
@idoshamun We just need to verify my JSON expected input matches what Denis is sending.

Introduces a new post-updated subscription, can update any part of the post
Can determine if post becomes visible
Updates shared post to become visible.

WT-1127 #done

Review suggestion:
Ensure test cases are solid.

rebelchris commented on .infra/workers.ts:
```diff
@@ -171,4 +171,8 @@ export const workers: Worker[] = [
   },
+  {
+    topic: 'yggdrasil.v1.content-published',
+    subscription: 'api.post-updated',
```
We can go with `api.content-published`?

idoshamun replied:
Yes that works :)

idoshamun commented on src/workers/postUpdated.ts:
```diff
@@ -0,0 +1,68 @@
+
+const worker: Worker = {
+  subscription: 'api.post-updated',
+  handler: async (message, con, logger): Promise<void> => {
```
This handler misses a use case for crawler and community picks where we don't prematurely generate an id that Yggdrasil can utilize. Am I right?

rebelchris replied:
Ah good point, not sure it's high impact for now, but we can add it.
So far we have that retro use case only for later down the line.

Let me know what you think.

rebelchris replied:
@idoshamun We need to think this trough, for community picks we need to change it regardless so the post will be created same as with squads now.
For crawler I agree it's an exception, but what rules will we set since it's not "updating" then but creating a completely new one.

idoshamun replied:
Please confirm with Viktor that we can really tackle it later. If so, that's what will do. Ask him on #engineering please

idoshamun commented on src/workers/postUpdated.ts:
```diff
@@ -0,0 +1,68 @@
+            sourceId: data?.sourceId,
+            image: data?.image,
+            title: data?.title,
+            metadataChangedAt: updatedDate,
```
One thing that I noticed is that we change this value ourselves in this CDC in some cases. 
I think we should remove some of this cases from CDC (for example when `visible` changes) because it addresses here and it may create race conditions.

rebelchris replied:
We can remove the other ones if that's what you mean.
Since those are not actually using it.

idoshamun replied:
Tinybird exporter uses it. I'll review it now again and let you know what props can be ignored in the cdc

idoshamun replied:
I believe we can keep only this lines:
isChanged(data.payload.before, data.payload.after, 'deleted') ||
      isChanged(data.payload.before, data.payload.after, 'banned') ||
      isChanged(data.payload.before, data.payload.after, 'tagsStr')

idoshamun commented on src/workers/postUpdated.ts:
```diff
@@ -0,0 +1,68 @@
+            description: data?.description,
+            readTime: data?.readTime,
+            toc: data?.toc,
+            keywords: data?.keywords,
```
There's a whole logic for the keywords update.
I suggest having a look at our existing `newPost` endpoint to make sure we don't miss anything.
Essentially, this worker will take charge after we finish the migration, so we need to ensure everything is here.

idoshamun reviewed (changes requested):
That's a good start but still misses some use cases I think

idoshamun commented on src/workers/postUpdated.ts:
```diff
@@ -0,0 +1,68 @@
+            keywords: data?.keywords,
+            siteTwitter: data?.siteTwitter,
+            creatorTwitter: data?.creatorTwitter,
+            sourceId: data?.sourceId,
```
We should probably update the `private` field based on the source

rebelchris commented:
@idoshamun Refactored so it uses the existing newPost flow.
Tried to re-use as much as possible, although there is still some overlap (As newPost consists of creation)

Love to hear what you think.

idoshamun commented on src/entity/posts/utils.ts:
```diff
@@ -257,11 +257,21 @@ const addPostAndKeywordsToDb = async (
+  postId: string,
+): Promise<void> => {
+  if (tags?.length) {
     await entityManager.getRepository(PostTag).insert(
```
By now this is no longer used and is actually deprecated. We use only the keywords

rebelchris replied:
Will remove, also will impact the existing new post then.

idoshamun commented on src/workers/postUpdated.ts:
```diff
@@ -0,0 +1,120 @@
+          visibleAt: becomesVisible ? updatedDate : null,
+          tagsStr: allowedKeywords?.join(',') || null,
+          private: privacy,
+          sentAnalyticsReport: !data?.authorId,
```
If the post is private we shouldn't send analytics report

```suggestion
          sentAnalyticsReport: private || !data?.authorId,
```

idoshamun commented on __tests__/workers/postUpdated.ts:
```diff
@@ -0,0 +1,107 @@
+    .getRepository(ArticlePost)
+    .update({ id: 'p1' }, { origin: PostOrigin.CommunityPicks });
+  await expectSuccessfulBackground(worker, {
+    updated_at: new Date('01-05-1990 12:00:00'),
```
An id is missing here that indicate what post to look at. And updated_at is too old compared to the metadata changed at

idoshamun commented on __tests__/workers/postUpdated.ts:
```diff
@@ -0,0 +1,107 @@
+
+it('should not update if the database updated date is newer', async () => {
+  await expectSuccessfulBackground(worker, {
+    updated_at: new Date('01-05-1990 12:00:00'),
```
No id provided here

idoshamun commented on src/workers/postUpdated.ts:
```diff
@@ -0,0 +1,120 @@
+  handler: async (message, con, logger): Promise<void> => {
+    const data: Data = messageToJson(message);
+    try {
+      const { id, tags, keywords, updated_at, ...submittedData } = data;
```
Please add a check that id is provided before you proceed

idoshamun reviewed (changes requested):
I noticed an id is missing in the payload of all your tests. Please also add applicable tests from the new post file. For example that we set proper tags and so on.
Sorry for the tight review, it's just an important part of the app

rebelchris commented:
@idoshamun I added the ID check and test case for empty ID.
Also removed the tags and added the relevant test.

It seems no other private newPost test would be applicable as we already check everything (except community picks which is not yet applicable)

idoshamun commented on __tests__/workers/postUpdated.ts:
```diff
@@ -0,0 +1,174 @@
+  expect(post.visible).toEqual(true);
+  expect(post.visibleAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
+  expect(post.title).toEqual('test');
+  const keywords = await con.getRepository(Keyword).find({
```
We are not checking completely the keywords here, let's add a more thorough check on `tagsStr`

idoshamun reviewed (changes requested):
One last comment
